substrate/primitives/runtime/src/nonce.rs

```diff
@@ -0,0 +1,498 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::traits::{Bounded, Nonce, One, Zero};
+use codec::{Compact, CompactAs, Decode, Encode, MaxEncodedLen};
+use core::{
+	fmt::Display,
+	marker::PhantomData,
+	ops::{
+		Add, AddAssign, BitAnd, BitOr, BitXor, Deref, Div, DivAssign, Mul, MulAssign, Not, Rem,
+		RemAssign, Shl, Shr, Sub, SubAssign,
+	},
+};
+use num_traits::{
+	CheckedAdd, CheckedDiv, CheckedMul, CheckedNeg, CheckedRem, CheckedShl, CheckedShr, CheckedSub,
+	Num, NumCast, PrimInt, Saturating, ToPrimitive,
+};
+use scale_info::TypeInfo;
+use sp_core::Get;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[derive(Encode, Decode, TypeInfo, Debug, MaxEncodedLen)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct NonceWithDefault<D: Get<N>, N: Nonce>(N, PhantomData<D>);
```

Comment:
```suggestion
pub struct TypeWithDefault<D: Get<N>, N: Nonce>(N, PhantomData<D>);
```

Or whatever. The name should be generic. No need to pin point this to `Nonce`

Reply:
But it does rely on `N` to implement a set of traits that provide arithmetic ops, etc.

Reply:
Oh, let me try another way then - probably implementing these traits individually when the underlying type implements them too.

Reply:
Added.